File: src/demo_python_tf/demo_python_tf/static_tf_broadcaster.py
import rclpy
from rclpy.node import Node
from tf2_ros import StaticTransformBroadcaster # 静态坐标变换广播器
from geometry_msgs.msg import TransformStamped # 消息接口
from tf_transformations import quaternion_from_euler # 欧拉角转四元数
import math # 角度转弧度

class StaticTFBroadcaster(Node):
    
    def __init__(self):
        super().__init__("static_tf_broadcaster")
        self.static_broadcaster_ = StaticTransformBroadcaster(self) # 创建静态坐标变换广播器
        self.publish_static_tf() # 发布静态坐标变换


    def publish_static_tf(self):
        """
        发布静态TF变换, 从base_link到camera_link
        """
        transform = TransformStamped() # 创建坐标变换消息
        transform.header.frame_id = "base_link" # 父坐标系
        transform.child_frame_id = "camera_link" # 子坐标系
        transform.header.stamp = self.get_clock().now().to_msg() # 时间戳

        transform.transform.translation.x = 0.5 # 平移
        transform.transform.translation.y = 0.3
        transform.transform.translation.z = 0.6

        # 欧拉角转四元数 q = x, y, z, w
        q = quaternion_from_euler(math.radians(180), 0, 0) # 角度转弧度
        # 设置旋转
        transform.transform.rotation.x = q[0]
        transform.transform.rotation.y = q[1]
        transform.transform.rotation.z = q[2]
        transform.transform.rotation.w = q[3]
        # 发布静态坐标变换
        self.static_broadcaster_.sendTransform(transform)
        self.get_logger().info(f"Published static transform {transform}")


def main():
    rclpy.init()
    node = StaticTFBroadcaster()
    rclpy.spin(node)
    rclpy.shutdown()

File: src/demo_cpp_tf/src/static_tf_broadcaster.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp" // 提供消息接口
#include "tf2/LinearMath/Quaternion.h" // 提供tf2::Quaternion类
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp" // 消息类型转换函数
#include "tf2_ros/static_transform_broadcaster.h" // 静态坐标广播器类


class StaticTFBroadcaster: public rclcpp::Node
{
private:
    std::shared_ptr<tf2_ros::StaticTransformBroadcaster> broadcaster_; // 坐标广播器对象指针
public:
    StaticTFBroadcaster() : Node("static_tf_broadcaster")
    {
        broadcaster_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);
        this->publish_tf();
    }

    void publish_tf()
    {
        geometry_msgs::msg::TransformStamped transform;

        transform.header.stamp = this->get_clock()->now(); // 设置时间戳
        transform.header.frame_id = "map"; // 设置父坐标系
        transform.child_frame_id = "target_point"; // 设置子坐标系

        transform.transform.translation.x = 5.0; // 设置平移分量
        transform.transform.translation.y = 3.0; // 设置平移分量
        transform.transform.translation.z = 0.0; // 设置平移分量

        tf2::Quaternion q; // 创建tf2::Quaternion对象
        q.setRPY(0.0, 0.0, 60*M_PI/180); // 设置欧拉角（滚转、俯仰、偏航）
        transform.transform.rotation = tf2::toMsg(q); // 设置旋转分量

        this->broadcaster_->sendTransform(transform); // 发布坐标变换
    }
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv); 
    auto node = std::make_shared<StaticTFBroadcaster>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

